From 9c1464707679de152568f6257079828c3c466265 Mon Sep 17 00:00:00 2001
From: David Llewellyn-Jones <david.llewellyn-jones@jolla.com>
Date: Fri, 7 Feb 2020 19:29:59 +0200
Subject: [PATCH] Set new IMAP folders to inherit SynchronizationEnabled flag

When new folders are received from the server, the client-side
SynchronizationEnabled flag must be set for them. Previously they were
always set to true, but if they're inside a folder that isn't being
synced, this probably doesn't reflect the user's intent.

This change makes it so that for IMAP any new folders received from the
server will have the SynchronizationEnabled flag set to whatever value
their parent has it set to. If they have no parent (i.e. are in the root
folder) then the flag is set to true as before.

Change-Id: If43d5e2b4469a402d203841c04cc4b1b30ffc3b1
---
 src/plugins/messageservices/imap/imapclient.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/plugins/messageservices/imap/imapclient.cpp b/src/plugins/messageservices/imap/imapclient.cpp
index d78d897e..cfb11519 100644
--- a/src/plugins/messageservices/imap/imapclient.cpp
+++ b/src/plugins/messageservices/imap/imapclient.cpp
@@ -950,8 +950,11 @@ void ImapClient::mailboxListed(const QString &flags, const QString &path)
             // This element needs to be created
             QMailFolder folder(mailboxPath, parentId, _config.id());
             folder.setDisplayName(QMailCodec::decodeModifiedUtf7(*it));
-            folder.setStatus(QMailFolder::SynchronizationEnabled, true);
             folder.setStatus(QMailFolder::Incoming, true);
+            // Set synchronization flag the same as parent folder, or true if there's no parent
+            bool synchronize = parentId.isValid() ?
+                        QMailFolder(parentId).status() & QMailFolder::SynchronizationEnabled : true;
+            folder.setStatus(QMailFolder::SynchronizationEnabled, synchronize);
 
             // The reported flags pertain to the listed folder only
             QString folderFlags;
-- 
2.26.2

